cmake/FindFuthark.cmake

```diff
@@ -0,0 +1,152 @@
+# TRACCC library, part of the ACTS project (R&D line)
+#
+# (c) 2022 CERN for the benefit of the ACTS project
+#
+# Mozilla Public License Version 2.0
+
+# Necessary CMake includes
+include(FindPackageHandleStandardArgs)
+include(CheckLanguage)
+
+# The C language is not enabled in our project by default, but we need it to
+# compile Futhark's code.
+check_language(C)
+if(CMAKE_C_COMPILER)
+    enable_language(C)
+endif()
+
+# Try to find our dependencies, which are of course Futhark itself (required)
+# and the CUDA toolkit for compiling CUDA projects. Also notify the user of
+# what we find.
+find_program(Futhark_EXECUTABLE futhark)
+mark_as_advanced(Futhark_EXECUTABLE)
+find_package(CUDAToolkit)
+
+# Report the status of loading Futhark to the user.
+find_package_handle_standard_args(
+    Futhark
+    REQUIRED_VARS
+    CMAKE_C_COMPILER
+    Futhark_EXECUTABLE
+)
+
+# This function describes how to add a Futhark source to a CMake library. The
+# function adds sources generated from a source file `_SOURCE_NAME` to a target
+# called `TARGET_NAME`. The Futhark target platform (CUDA or C) is given
+# (case-insensitively) by `_LANGUAGE_TARGET`
+function(add_futhark_to_library TARGET_NAME _LANGUAGE_TARGET _SOURCE_NAME)
+    # Parse the arguments that we get.
+    cmake_parse_arguments(
+        ARGS
+        ""
+        ""
+        "DEPENDENCIES"
+        ${ARGN}
+    )
+
+    # Set a few useful variables.
+    set(FUTHARK_LANGUAGES c cuda)
+    set(SOURCE_NAME ${CMAKE_CURRENT_SOURCE_DIR}/${_SOURCE_NAME})
+    string(TOLOWER ${_LANGUAGE_TARGET} LANGUAGE_TARGET)
+    string(TOUPPER ${_LANGUAGE_TARGET} LANGUAGE_TARGET_FANCY)
+
+    # Notify the user of what we are doing.
+    message(
+        STATUS
+        "Building ${LANGUAGE_TARGET_FANCY} Futhark target ${TARGET_NAME} from ${SOURCE_NAME}"
+    )
+
+    # Make sure that the source file actually exists.
+    if(NOT EXISTS ${SOURCE_NAME})
+        message(
+            FATAL_ERROR
+            "File ${SOURCE_NAME} does not exist!"
+        )
+    endif()
+
+    # Ensure that the requested language is actually a supported one, namely C
+    # or CUDA.
+    list(FIND FUTHARK_LANGUAGES ${LANGUAGE_TARGET} index)
+    if(index EQUAL -1)
+        message(
+            FATAL_ERROR
+            "Futhark language must be one of ${FUTHARK_LANGUAGES}, not ${LANGUAGE_TARGET}"
+        )
+    endif()
+
+    # Define some needed directory and file names.
+    get_filename_component(FILE_ROOT ${SOURCE_NAME} NAME_WE)
+    set(FUTHARK_OUTPUT_SOURCE_DIR ${CMAKE_CURRENT_BINARY_DIR}/src)
+    set(FUTHARK_OUTPUT_HEADER_DIR ${CMAKE_CURRENT_BINARY_DIR}/include/traccc/futhark)
```

Comment:
Let's make everything be under `"${CMAKE_CURRENT_BINARY_DIR}${CMAKE_FILES_DIRECTORY}/"`. `CMAKE_FILES_DIRECTORY` is a very badly documented variable that pretty much is just an alias for `"/CMakeFiles.dir"`. (With the leading slash.)

The point is, with an "in source" build such a setup makes sure that all the generated files would be tucked away in the directory anyway created by CMake. And not possibly interfere with the source files in the source directory.